includes/admin.form.inc

```diff
@@ -127,23 +132,41 @@ function islandora_video_admin_submit($form, &$form_state) {
  * Form validation ensuring the selected ffmpeg options are available.
  */
 function islandora_video_admin_validate($form, &$form_state) {
-  // Ensure the mp4 audio codec is enabled.
-  $raw_value = $form_state['values']['islandora_video_mp4_audio_codec'];
-  if ($raw_value) {
-    if (preg_match('/[^0-9a-zA-Z_-]/', $raw_value) === 1) {
-      form_set_error('', "The value entered for MP4 audio codec contains forbidden characters.");
-      return;
+  // Ensure the mp4 audio codec is present if MP4 derivative is enabled.
+  if ($form_state['values']['islandora_video_make_mp4_locally'] == TRUE) {
+    $raw_value = $form_state['values']['islandora_video_mp4_audio_codec'];
+    if ($raw_value) {
+      if (preg_match('/[^0-9a-zA-Z_-]/', $raw_value) === 1) {
+        form_set_error('islandora_video_mp4_audio_codec', "The value entered for MP4 audio codec contains forbidden characters.");
+        return;
+      }
+      // Lower-case the value because all codecs in ffmpeg are lower case.
+      $codec = strtolower(trim($raw_value));
+      $form_state['values']['islandora_video_mp4_audio_codec'] = $codec;
+
+      // Test that FFMPEG path is valid.
+      $ffmpeg = ($form_state['values']['islandora_video_ffmpeg_path'] !== variable_get('islandora_video_ffmpeg_path', 'ffmpeg') ? $form_state['values']['islandora_video_ffmpeg_path'] : variable_get('islandora_video_ffmpeg_path', 'ffmpeg'));
+      if (preg_match('/[^0-9a-zA-Z\\/\\\\_-]/', $ffmpeg) === 1) {
+        form_set_error('islandora_video_ffmpeg_path', "The value entered for FFmpeg path contains forbidden characters.");
+        return;
+      }
+      $command = $ffmpeg . ' -version';
+      exec($command, $output, $ret);
+      if ($ret !== 0) {
+        form_set_error('islandora_video_ffmpeg_path', "FFmpeg was not found. A valid FFmpeg path is required if MP4 derivative generation is enabled.");
+      }
+      else {
+        // Test that the specified codec is present.
+        unset($output);
+        $command = $ffmpeg . ' -encoders 2>/dev/null | grep "^ ...... ' . $codec . ' "';
+        exec($command, $output, $ret);
+        if (!$output) {
+          form_set_error('islandora_video_mp4_audio_codec', 'The selected MP4 codec was not found in ffmpeg. Try using aac or enable the desired codec.');
+        }
+      }
     }
-    $ffmpeg = ($form_state['values']['islandora_video_ffmpeg_path'] !== variable_get('islandora_video_ffmpeg_path', 'ffmpeg') ? $form_state['values']['islandora_video_ffmpeg_path'] : variable_get('islandora_video_ffmpeg_path', 'ffmpeg'));
-    $safe_value = strtolower(trim($raw_value));
-    $command = $ffmpeg . ' -encoders 2>/dev/null | grep "^ ...... ' . $safe_value . ' "';
-    exec($command, $output, $ret);
-    if (!$output) {
-      form_set_error('', 'The selected MP4 codec was not found in ffmpeg. Try using aac or enable the desired codec.');
+    else {
+      $form_state['values']['islandora_video_mp4_audio_codec'] = 'libfaac';
```

Comment:
Wait, I'm confused. What should happen when someone empties the codec field and saves? In the other comment, online 144, you said I should validate ffmpeg, if someone empties the codec field. But the form's not truly valid with MP4 enabled, ffmpeg ok, and no valid codec, so I made them use libfaac (the default). Doesn't that make sense? If you clear out a value and try to save, I think it makes sense for the default to appear rather than the previously saved value.

Reply:
@rosiel but on a sane installation, you can't save if the codec is not valid right? So whatever you have saved should be OK(or could be ok, an exception is FFmpeg was changed, that is why we need to test for ffmpeg always) and should have priority over the islandora default. And if that does not exist, it will fall back to whatever default we provide. I mean, we can tell people, hey you removed all values, we give you back what you had before. But imagine someone has "rosieaac", by mistake removes the value and tries to save, and now that person gets libfaac there, even when rosieaac was the right one (and could happen this user don't even notice the change and saves). Or maybe I get this wrong. @whikloj you use Video too right?

Reply:
It does seem a little odd still.  Maybe since the field gets validated, a blank should be immediately invalid - just return the form to the user with their blank and a message saying "this can't be blank". That seems better than either assuming "you must have meant libfaac, let's validate and save that" or "you must have meant to not change anything, let's validate and save the previously stored value." (I have code to this effect,  here https://github.com/rosiel/islandora_solution_pack_video/compare/7.x-ISLANDORA-2160...rosiel:7.x-ISLANDORA-2160a)